### DISCUSSION TITLE
Save each layer in a DXF into a seperate pdf or dxf file

### DISCUSSION THREAD

**Discussion by vikm2o** [mozman/ezdxf#756] [Q&A] [answered]
I have a DXF file with sewing patterns. There are multiple layers for each size for e.g XXS, XS,....XXL. I want to save each layer as a separate DFX or pdf file.

I am able to query the layers using ```doc.modelspace().query('*[layer=="PDF_XLARGE"]'```
but I am not able to save it to a separate dxf file.

I tried creating a new doc and adding the entities returned by the query, but I get this error.
```DXFStructureError: Adding entities from a different DXF drawing is not supported.```

I am a complete beginner, and I am not sure how to proceed.

**Comment by mozman** [comment, marked as answer]
I added an  [example](https://github.com/mozman/ezdxf/blob/037816b943bf067dcdd9727777965fab5f057e54/examples/separate_layers.py) for that task.

```Python
def main(filename: str):
    source_doc = ezdxf.readfile(CWD / filename)
    source_msp = source_doc.modelspace()
    # create an EntityQuery container with all entities from the modelspace
    source_entities = source_msp.query()
    # get all layer names from entities in the modelspace:
    all_layer_names = [e.dxf.layer for e in source_entities]
    # remove unwanted layers if needed

    for layer_name in all_layer_names:
        # create a new document for each layer with the same DXF version as the
        # source file
        layer_doc = ezdxf.new(dxfversion=source_doc.dxfversion)
        layer_msp = layer_doc.modelspace()
        importer = Importer(source_doc, layer_doc)
        # select all entities from modelspace from this layer (extended query
        # feature of class EntityQuery)
        entities: EntityQuery = source_entities.layer == layer_name  # type: ignore
        if len(entities):
            importer.import_entities(entities, layer_msp)
            # create required resources
            importer.finalize()
            layer_doc.saveas(CWD / f"{layer_name}.dxf")

```
The example imports all entities from the modelspace into a new DXF document. The Importer add-on is required to also import the required resources used by the imported entities.

Important advice:
Keep things simple, more complex DXF documents may fail or at least the entities will not look the same as in the source file! Read the docs!

Basic concept of layers : https://ezdxf.mozman.at/docs/concepts/layers.html
Importer add-on: https://ezdxf.mozman.at/docs/addons/index.html

**Reply by vikm2o** [reply]
Thanks a lot ! Works really well!

**Reply by vikm2o** [reply]
The output DXF appears to be much smaller when opened in AutoCAD. I added the below lines to the import layout.
```
# import all paperspace layouts from source drawing
importer.import_paperspace_layouts()
```
It didn't make any difference.
